src/lib/discovery.js

```diff
@@ -11,78 +11,159 @@ import EventEmitter from 'eventemitter3'
 import Client from './client'
 
 export class SKServer {
-  constructor (service) {
+  constructor(service) {
     this._roles = service.roles || ['master', 'main']
     this._self = service.self || ''
     this._version = service.version || '0.0.0'
     this._hostname = service.hostname
     this._port = service.port
   }
 
-  get roles () {
+  get roles() {
     return this._roles
   }
 
-  get self () {
+  get self() {
     return this._self
   }
 
-  get version () {
+  get version() {
     return this._version
   }
 
-  get hostname () {
+  get hostname() {
     return this._hostname
   }
 
-  get port () {
+  get port() {
     return this._port
   }
 
-  isMain () {
+  isMain() {
     return this._roles.includes('main')
   }
 
-  isMaster () {
+  isMaster() {
     return this._roles.includes('master')
   }
 
-  createClient (opts = {}) {
+  createClient(opts = {}) {
     return new Client({
       ...opts,
       hostname: this._hostname,
-      port: this._port
+      port: this._port,
     })
   }
 }
 
 export default class Discovery extends EventEmitter {
-  constructor (bonjour, timeout = 60000) {
+  constructor(bonjourOrMdns, timeout = 60000) {
     super()
 
-    const props = [ '_server', '_registry' ].join(',')
+    this.found = []
 
-    if (!bonjour || typeof bonjour !== 'object' || Object.keys(bonjour).join(',') !== props) {
-      throw new Error('Invalid mDNS provider')
+    if (!bonjourOrMdns || typeof bonjourOrMdns !== 'object') {
+      throw new Error('No mDNS provider given')
     }
 
-    this.found = []
+    const bonjourProps = ['_server', '_registry'].join(',')
+    const mdnsProps = ['dns_sd', 'Advertisement', 'createAdvertisement', 'Browser'].join(',')
+
+    if (Object.keys(bonjourOrMdns).join(',').startsWith(bonjourProps)) {
+      return this.discoverWithBonjour(bonjourOrMdns, timeout)
+    }
+
+    if (Object.keys(bonjourOrMdns).join(',').startsWith(mdnsProps)) {
+      return this.discoverWithMdns(bonjourOrMdns, timeout)
+    }
+
+    throw new Error('Unrecognized mDNS provider given')
+  }
+
+  discoverWithBonjour(bonjour, timeout) {
     const browser = bonjour.find({ type: 'signalk-http' })
 
-    browser.on('up', ad => {
+    browser.on('up', (ad) => {
       const service = {
         ...ad.txt,
         name: ad.name || '',
         hostname: ad.host || '',
-        port: parseInt(ad.port, 10)
+        port: parseInt(ad.port, 10),
+        provider: 'bonjour',
+      }
+
+      if (
+        service.hasOwnProperty('roles') &&
+        typeof service.roles === 'string' &&
+        service.roles.includes(',')
+      ) {
+        service.roles = service.roles.split(',').map((role) => role.trim().toLowerCase())
+      }
+
+      if (
+        service.hasOwnProperty('roles') &&
+        typeof service.roles === 'string' &&
+        !service.roles.includes(',')
+      ) {
+        service.roles = [service.roles].map((role) => role.trim().toLowerCase())
+      }
```

Comment:
Deduplicate these between the discovery methods.

The code seems complex to me, hard to understand what's going on. Why the includes(',') test, as `'foo'.split(',')` yields an array anyway?

Reply:
Good catch, I reckon it was pretty late when I wrote that. Very nicely overengineered and overcomplicated.